sdlib/d/gc/tcache.d

```diff
@@ -638,7 +657,10 @@ private:
 
 	void clearTLSSegments() {
 		if (tlsSegments.ptr !is null) {
-			free(tlsSegments.ptr);
+			// Do not call free directly, because that might try to
+			// change the cache bins, and this can be called from a
+			// signal handler.
+			freeToArena(tlsSegments.ptr);
```

Comment:
I would look at what call site this has. Clearing the tls segment doesn't in itself require this change, so that seems misplaced. this change is only required when done from a signal handler.

But I strongly suspect we only do so from a signal handler anyways. Maybe on thread shutdown, but in this case, we don't need to check for the point, as free(null) works and we don't need to update the value as the thread is going away anyways, so that could just be a free.

If that is correct, this the proper thing to do here is to move that logic, alongside with the free to arena thing in the pre suspend hook.

Reply:
Rearranged. I think the logic does fit better. And because the other place this is called is outside the signal handler (on thread shutdown), we can just call the normal free at that point.